README.md
# Emysql

[![Build Status](https://travis-ci.org/qingchuwudi/Emysql.svg?branch=master)](https://travis-ci.org/qingchuwudi/Emysql)

This is from [Eonblast/Emysql](https://github.com/Eonblast/Emysql)

To use this driver, see the [Samples] further down in this README document.

## Introduction

[Eonblast/Emysql/README](https://github.com/Eonblast/Emysql/README.md)

[inaka/Emysql/README](https://github.com/inaka/Emysql/README.md)

<hr/>

- **Which fork/package should I use?** Likely this one, but see [Choosing][].

- **Why are there so many?** See [History][].

- **Who used *this* fork?** [Electronic Arts][History].

- **How do I ...?** See [Samples][].

- **Hello ...?** See [Samples][].

- **Download:** <https://github.com/Eonblast/Emysql/archives/master>

- **Docs:** <http://eonblast.github.com/Emysql/>

- **Issues:** <https://github.com/Eonblast/Emysql/issues>

- **Repository:** <https://github.com/Eonblast/Emysql>

<hr/>

## Contents

* [Choosing][]
* [Samples][]
* [Usage][]
* [Tests][]
* [History][]
* [Links][]
* [Todo][]
* [License][]

<hr/>

## Choosing                                             <a name="Choosing"></a>

#### Best
In most cases, especially for high performance and stability, this package, [Emysql][emysql], will be the best choice. It was rewritten from the ground up to overcome fundamental issues of 'erlang-mysql-driver'. It also has some usable docs meanwhile.

#### Simple
If you are looking for the **plain necessities**, you should use the [ejabberd][7] mysql driver. It is simple, battle tested and stable. There are comprehensive instructions in the source comments.

#### Transaction

This driver currently does not support transactions.

For **mnesia-style transactions**, one of the multiple '[erlang-mysql-driver][22]s' may suite you best. There are [quite many][16] branches of it out there, and they are based on the same project as the ejabberd driver. To learn more about out the differences between the drivers, see the [mysql driver history][History].

## Samples                                             <a name=Samples></a>

### Hello World

This is a hello world program. Follow the three steps below to try it out.

```Erlang
-module(a_hello).
-export([run/0]).

run() ->

        crypto:start(),
        application:start(emysql),

        emysql:add_pool(hello_pool, [{size,1},
                     {user,"hello_username"},
                     {password,"hello_password"},
                     {database,"hello_database"},
                     {encoding,utf8}]),

        emysql:execute(hello_pool,
                <<"INSERT INTO hello_table SET hello_text = 'Hello World!'">>),

        Result = emysql:execute(hello_pool,
                <<"select hello_text from hello_table">>),

        io:format("~n~p~n", [Result]).
```

We'll be coming back to this source to make it run on your machine in a minute. But let's look at the basic building blocks first:

### Executing an SQL Statement

```Erlang
emysql:execute(my_pool, <<"SELECT * from mytable">>).
```

For the exact spec, see below, [Usage][]. Regarding the 'pool', also see below.

### Executing a Prepared Statement

```Erlang
emysql:prepare(my_stmt, <<"SELECT * from mytable WHERE id = ?">>).

emysql:execute(my_pool, my_stmt, [1]).
```

### Executing Stored Procedures

```Erlang
emysql:execute(my_pool, <<"create procedure my_sp() begin select * from mytable; end">>).

emysql:execute(my_pool, <<"call my_sp();">>).
```

### Result Record

```Erlang
-record(result_packet, {seq_num, field_list, rows, extra}).
```
	
### Converting Row Data To Records

```Erlang
-record(foo, {bar, baz}).

Result = emysql:execute(pool1, <<"select bar, baz from foo">>).
Recs = emysql:as_record(Result, foo, record_info(fields, foo)).
Bars = [Foo#foo.bar || Foo <- Recs].
```

### Adding a Connection to the Connection Pool

Emysql uses a sophisticated connection pooling mechanism.

```Erlang
emysql:add_pool(my_pool, [{size,1}, {user,"myuser"}, {password,"mypass"},
        {host,"myhost"}, {port,3306},{database,"mydatabase"}, {encoding,utf8}]).
```

Arbitrary post-connection start-up commands can be added with the proplist key atom
`start_cmds`:

```Erlang
emysql:add_pool(my_pool, [{size,1}, {user,"myuser"}, {password,"mypass"},
        {host,"myhost"}, {port,3306},{database,"mydatabase"}, {encoding,utf8},
        {start_cmds,[
                <<"SET TIME_ZONE='+00:00'">>,
                <<"SET SQL_MODE='STRICT_ALL_TABLES'">>
        ]}]).
```

### Running Hello World

Let's run the hello world sample from above:

#### 1. Build Emysql

Build emysql.app, using make:

	$ cd Emysql
	$ make

Or use rebar:

	$ cd Emysql
	$ ./rebar compile

Both yield an option to install but this is not needed for the samples.

#### 2. Make a Sample Database

For use in the above sample (and all of those below, too), create a local mysql database. You should have a mysql server installed and running:
	
	$ mysql [-u<user> -p]
	mysql> create database hello_database;
	mysql> use hello_database;
	mysql> create table hello_table (hello_text char(20));
	mysql> grant all privileges on hello_database.* to hello_username@localhost identified by 'hello_password';

#### 3. Paste & Run Hello

Be sure to have ./ebin in your Erlang path. The hello-world source as shown above already waits in the Emysql directory, as hello.erl. Just compile and run it:

	$ erlc hello.erl
	$ erl -pa ./ebin -s hello run -s init stop -noshell

That's it. If you need to blindly repeat that more often some time, you can also use

	$ make hello

There are more sample programs:

## More Samples
Sample programs are in ./samples.

* [a_hello](http://github.com/Eonblast/Emysql/blob/master/samples/a_hello.erl) - Hello World
* [b_raw](http://github.com/Eonblast/Emysql/blob/master/samples/b_raw.erl) - Hello World, raw output
* [c\_rows\_as\_records](http://github.com/Eonblast/Emysql/blob/master/samples/c_rows_as_records.erl) - Using Erlang records to access result rows
* [d\_prepared\_statement](http://github.com/Eonblast/Emysql/blob/master/samples/d_prepared_statement.erl) - Using prepared statements
* [e\_stored\_procedure](http://github.com/Eonblast/Emysql/blob/master/samples/e_stored_procedure.erl) - Using stored procedures

CHANGELOG.md
# Change Log

## [0.5.0](https://github.com/inaka/Emysql/tree/0.5.0) (2018-07-16)
[Full Changelog](https://github.com/inaka/Emysql/compare/0.4.2...0.5.0)

**Closed issues:**

- Version Bump to 0.4.4 [\#9](https://github.com/inaka/Emysql/issues/9)
- Version Bump to 0.4.3 [\#6](https://github.com/inaka/Emysql/issues/6)
- Move from erlang.mk to rebar3 [\#3](https://github.com/inaka/Emysql/issues/3)

**Merged pull requests:**

- Update from @Qingchuwudi's master [\#17](https://github.com/inaka/Emysql/pull/17) ([elbrujohalcon](https://github.com/elbrujohalcon))
- Fix connection leak when add\_pool is called in rapid succession [\#15](https://github.com/inaka/Emysql/pull/15) ([huss01](https://github.com/huss01))
- Fix dialyzer warnings. [\#13](https://github.com/inaka/Emysql/pull/13) ([f355](https://github.com/f355))
- Revert "\[Close \#9\] version bump to 0.4.3" [\#12](https://github.com/inaka/Emysql/pull/12) ([Euen](https://github.com/Euen))
- \[Close \#9\] version bump to 0.4.3 [\#11](https://github.com/inaka/Emysql/pull/11) ([Euen](https://github.com/Euen))
- \[Close \#9\] version bump to 0.4.4 [\#10](https://github.com/inaka/Emysql/pull/10) ([Euen](https://github.com/Euen))
- fix rebar3 compilation [\#8](https://github.com/inaka/Emysql/pull/8) ([Euen](https://github.com/Euen))
- \[Close \#6\] version bump to 0.4.3 [\#7](https://github.com/inaka/Emysql/pull/7) ([Euen](https://github.com/Euen))
- Fix bug when import library through hex.pm [\#5](https://github.com/inaka/Emysql/pull/5) ([Euen](https://github.com/Euen))
- \[Close \#3\] update app.src to be published in hex.pm [\#4](https://github.com/inaka/Emysql/pull/4) ([Euen](https://github.com/Euen))

## [0.4.2](https://github.com/inaka/Emysql/tree/0.4.2) (2015-08-26)
[Full Changelog](https://github.com/inaka/Emysql/compare/v0.4.1...0.4.2)

**Merged pull requests:**

- Version Bump to 0.4.2 [\#2](https://github.com/inaka/Emysql/pull/2) ([elbrujohalcon](https://github.com/elbrujohalcon))
- Ready for Erlang 18.0 [\#1](https://github.com/inaka/Emysql/pull/1) ([elbrujohalcon](https://github.com/elbrujohalcon))

## [v0.4.1](https://github.com/inaka/Emysql/tree/v0.4.1) (2014-02-27)
[Full Changelog](https://github.com/inaka/Emysql/compare/0.4.1...v0.4.1)

## [0.4.1](https://github.com/inaka/Emysql/tree/0.4.1) (2014-02-27)
[Full Changelog](https://github.com/inaka/Emysql/compare/0.4.0...0.4.1)

## [0.4.0](https://github.com/inaka/Emysql/tree/0.4.0) (2014-02-26)
[Full Changelog](https://github.com/inaka/Emysql/compare/0.3.2...0.4.0)

## [0.3.2](https://github.com/inaka/Emysql/tree/0.3.2) (2014-02-24)
[Full Changelog](https://github.com/inaka/Emysql/compare/0.3.1...0.3.2)

## [0.3.1](https://github.com/inaka/Emysql/tree/0.3.1) (2014-02-24)
[Full Changelog](https://github.com/inaka/Emysql/compare/0.2.16...0.3.1)

## [0.2.16](https://github.com/inaka/Emysql/tree/0.2.16) (2013-06-21)
[Full Changelog](https://github.com/inaka/Emysql/compare/0.2.13...0.2.16)

## [0.2.13](https://github.com/inaka/Emysql/tree/0.2.13) (2012-03-26)



src/emysql_conn.erl
%% Copyright (c) 2009-2012
%% Bill Warnecke <bill@rupture.com>,
%% Jacob Vorreuter <jacob.vorreuter@gmail.com>,
%% Henning Diedrich <hd2010@eonblast.com>,
%% Eonblast Corporation <http://www.eonblast.com>
%%
%% Permission is  hereby  granted,  free of charge,  to any person
%% obtaining  a copy of this software and associated documentation
%% files (the "Software"),to deal in the Software without restric-
%% tion,  including  without  limitation the rights to use,  copy,
%% modify, merge,  publish,  distribute,  sublicense,  and/or sell
%% copies  of the  Software,  and to  permit  persons to  whom the
%% Software  is  furnished  to do  so,  subject  to the  following
%% conditions:
%%
%% The above  copyright notice and this permission notice shall be
%% included in all copies or substantial portions of the Software.
%%
%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
%% EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
%% OF  MERCHANTABILITY,  FITNESS  FOR  A  PARTICULAR  PURPOSE  AND
%% NONINFRINGEMENT. IN  NO  EVENT  SHALL  THE AUTHORS OR COPYRIGHT
%% HOLDERS  BE  LIABLE FOR  ANY CLAIM, DAMAGES OR OTHER LIABILITY,
%% WHETHER IN AN ACTION OF CONTRACT,  TORT  OR OTHERWISE,  ARISING
%% FROM,  OUT OF OR IN CONNECTION WITH THE SOFTWARE  OR THE USE OR
%% OTHER DEALINGS IN THE SOFTWARE.
%% @private
-module(emysql_conn).
-export([set_database/2, set_encoding/2,
        execute/3, transaction/2, prepare/3, unprepare/2,
        open_connections/1, open_connection/1,
        reset_connection/4, close_connections/1, close_connection/1,
        open_n_connections/3, hstate/1,
        test_connection/3, need_test_connection/1
]).

-include("emysql.hrl").

%% MYSQL COMMANDS
-define(COM_SLEEP, 16#00).
-define(COM_QUIT, 16#01).
-define(COM_INIT_DB, 16#02).
-define(COM_QUERY, 16#03).
-define(COM_FIELD_LIST, 16#04).
-define(COM_CREATE_DB, 16#05).
-define(COM_DROP_DB, 16#06).
-define(COM_REFRESH, 16#07).
-define(COM_SHUTDOWN, 16#08).
-define(COM_STATISTICS, 16#09).
-define(COM_PROCESS_INFO, 16#0a).
-define(COM_CONNECT, 16#0b).
-define(COM_PROCESS_KILL, 16#0c).
-define(COM_DEBUG, 16#0d).
-define(COM_PING, 16#0e).
-define(COM_TIME, 16#0f).
-define(COM_DELAYED_INSERT, 16#10).
-define(COM_CHANGE_USER, 16#11).
-define(COM_BINLOG_DUMP, 16#12).
-define(COM_TABLE_DUMP, 16#13).
-define(COM_CONNECT_OUT, 16#14).
-define(COM_REGISTER_SLAVE, 16#15).
-define(COM_STMT_PREPARE, 16#16).
-define(COM_STMT_EXECUTE, 16#17).
-define(COM_STMT_SEND_LONG_DATA, 16#18).
-define(COM_STMT_CLOSE, 16#19).
-define(COM_STMT_RESET, 16#1a).
-define(COM_SET_OPTION, 16#1b).
-define(COM_STMT_FETCH, 16#1c).

set_database(_, undefined) -> ok;
set_database(_, Empty) when Empty == ""; Empty == <<>> -> ok;
set_database(Connection, Database) ->
    Packet = <<?COM_QUERY, "use `", (iolist_to_binary(Database))/binary, "`">>,  % todo: utf8?
    emysql_tcp:send_and_recv_packet(Connection#emysql_connection.socket, Packet, 0).

set_encoding(_, undefined) -> ok;
set_encoding(Connection, {Encoding, Collation}) ->
    Packet = <<?COM_QUERY, "set names '", (erlang:atom_to_binary(Encoding, utf8))/binary, 
        "' collate '", (erlang:atom_to_binary(Collation, utf8))/binary,"'">>,
    emysql_tcp:send_and_recv_packet(Connection#emysql_connection.socket, Packet, 0);
set_encoding(Connection, Encoding) ->
    Packet = <<?COM_QUERY, "set names '", (erlang:atom_to_binary(Encoding, utf8))/binary, "'">>,
    emysql_tcp:send_and_recv_packet(Connection#emysql_connection.socket, Packet, 0).

%% @todo This can go away once the underlying socket accepts IOData
canonicalize_query(Q) when is_binary(Q) -> Q;
canonicalize_query(QL) when is_list(QL) -> iolist_to_binary(QL).

execute(Connection, StmtName, []) when is_atom(StmtName) ->
    prepare_statement(Connection, StmtName),
    StmtNameBin = atom_to_binary(StmtName, utf8),
    Packet = <<?COM_QUERY, "EXECUTE ", StmtNameBin/binary>>,
    send_recv(Connection, Packet);
execute(Connection, Query, []) ->
    QB = canonicalize_query(Query),
    Packet = <<?COM_QUERY, QB/binary>>,
    send_recv(Connection, Packet);
execute(Connection, Query, Args) when (is_list(Query) orelse is_binary(Query)) andalso is_list(Args) ->
    StmtName = "stmt_"++integer_to_list(erlang:phash2(Query)),
    ok = prepare(Connection, StmtName, Query),
    Ret =
    case set_params(Connection, 1, Args, undefined) of
        OK when is_record(OK, ok_packet) ->
            ParamNamesBin = list_to_binary(string:join([[$@ | integer_to_list(I)] || I <- lists:seq(1, length(Args))], ", ")),  % todo: utf8?
            Packet = <<?COM_QUERY, "EXECUTE ", (list_to_binary(StmtName))/binary, " USING ", ParamNamesBin/binary>>,  % todo: utf8?
            send_recv(Connection, Packet);
        Error ->
            Error
    end,
    unprepare(Connection, StmtName),
    Ret;

execute(Connection, StmtName, Args) when is_atom(StmtName), is_list(Args) ->
    prepare_statement(Connection, StmtName),
    case set_params(Connection, 1, Args, undefined) of
        OK when is_record(OK, ok_packet) ->
            ParamNamesBin = list_to_binary(string:join([[$@ | integer_to_list(I)] || I <- lists:seq(1, length(Args))], ", ")),  % todo: utf8?
            StmtNameBin = atom_to_binary(StmtName, utf8),
            Packet = <<?COM_QUERY, "EXECUTE ", StmtNameBin/binary, " USING ", ParamNamesBin/binary>>,
            send_recv(Connection, Packet);
        Error ->
            Error
    end.


%%@doc transaction
-spec transaction(Connection, Fun) -> Result when
        Connection :: #emysql_connection{},
        Fun        :: function(),
        Result     :: any().
transaction(Connection, Fun) when is_function(Fun, 1) ->
    case begin_transaction(Connection) of
        #ok_packet{} ->
            try Fun(Connection) of
                Val ->
                    case commit_transaction(Connection) of
                        #ok_packet{} ->
                            {atomic, Val};
                        #error_packet{} = ErrorPacket ->
                            {aborted, {commit_error, ErrorPacket}}
                    end
            catch
                _What:Exception ->
                    rollback_transaction(Connection),
                    case Exception of
                        {aborted, Reason} ->
                            {aborted, Reason};
                        _ ->
                            exit(Exception)
                    end
            end;
        #error_packet{} = ErrorPacket ->
            {aborted, {begin_error, ErrorPacket}}
    end.

begin_transaction(Connection) ->
    Packet = <<?COM_QUERY, "BEGIN">>,
    emysql_tcp:send_and_recv_packet(Connection#emysql_connection.socket, Packet, 0).

rollback_transaction(Connection) ->
    Packet = <<?COM_QUERY, "ROLLBACK">>,
    emysql_tcp:send_and_recv_packet(Connection#emysql_connection.socket, Packet, 0).

commit_transaction(Connection) ->
    Packet = <<?COM_QUERY, "COMMIT">>,
    emysql_tcp:send_and_recv_packet(Connection#emysql_connection.socket, Packet, 0).



prepare(Connection, Name, Statement) when is_atom(Name) ->
    prepare(Connection, atom_to_list(Name), Statement);
prepare(Connection, Name, Statement) ->
    StatementBin = encode(Statement, binary),
    Packet = <<?COM_QUERY, "PREPARE ", (list_to_binary(Name))/binary, " FROM ", StatementBin/binary>>,  % todo: utf8?
    case send_recv(Connection, Packet) of
        OK when is_record(OK, ok_packet) ->
            ok;
        Err when is_record(Err, error_packet) ->
            exit({failed_to_prepare_statement, Err#error_packet.msg})
    end.

unprepare(Connection, Name) when is_atom(Name)->
    unprepare(Connection, atom_to_list(Name));
unprepare(Connection, Name) ->
    Packet = <<?COM_QUERY, "DEALLOCATE PREPARE ", (list_to_binary(Name))/binary>>,  % todo: utf8?
    send_recv(Connection, Packet).

open_n_connections(PoolServer, PoolId, N) ->
    case emysql_conn_mgr:find_pool(PoolId, emysql_conn_mgr:pools(PoolServer)) of
        {Pool, _} ->
            lists:foldl(fun(_, {Conns, Reasons}) ->
                %% Catch {'EXIT',_} errors so newly opened connections are not orphaned.
                %% We do not want to close all the connections here like in
                %% open_connections/2. Struggle to keep working.
                case catch open_connection(Pool) of
                    #emysql_connection{} = Connection ->
                        {[Connection | Conns], Reasons};
                    {'EXIT', Reason} ->
                        {Conns, [Reason | Reasons]}
                end
            end, {[], []}, lists:seq(1, N));
        _ ->
            exit(pool_not_found)
    end.

%% @doc Opens connections for the necessary pool.
%%
%% If connection opening fails, removes all connections from the pool
%% Does not remove pool from emysql_conn_mgr due to a possible deadlock.
%% Caller must do it by itself.
open_connections(Pool) ->
     %-% io:format("open connections loop: .. "),
    case (queue:len(Pool#pool.available) + gb_trees:size(Pool#pool.locked)) < Pool#pool.size of
        true ->
            case catch open_connection(Pool) of
                #emysql_connection{} = Conn ->
                    open_connections(Pool#pool{available = queue:in(Conn, Pool#pool.available)});
                {'EXIT', Reason} ->
                    close_connections(Pool),
                    {error, Reason}
			end;
        false ->
            {ok, Pool}
    end.

open_connection(#pool{pool_id=PoolId, host=Host, port=Port, user=User,
        password=Password, database=Database, encoding=Encoding,
                      start_cmds=StartCmds, connect_timeout=ConnectTimeout} = Pool) ->
     %-% io:format("~p open connection for pool ~p host ~p port ~p user ~p base ~p~n", [self(), PoolId, Host, Port, User, Database]),
     %-% io:format("~p open connection: ... connect ... ~n", [self()]),
    case gen_tcp:connect(Host, Port, [binary, {packet, raw}, {active, false}, {recbuf, ?TCP_RECV_BUFFER}], ConnectTimeout) of
        {ok, Sock} ->
            #greeting {
               server_version = Version,
               thread_id = ThreadId,
               caps = Caps,
               language = Language
              } = handshake(Sock, User, Password),
            Connection = #emysql_connection{
                            id = erlang:port_to_list(Sock),
                            pool_id = PoolId,
                            encoding = Encoding,
                            socket = Sock,
                            version = Version,
                            thread_id = ThreadId,
                            caps = Caps,
                            language = Language,
                            test_period = Pool#pool.conn_test_period,
                            last_test_time = now_seconds(),
                            warnings = Pool#pool.warnings
                           },
            %%-% io:format("~p open connection: ... set db ...~n", [self()]),
            ok = set_database_or_die(Connection, Database),
            ok = set_encoding_or_die(Connection, Encoding),
            ok = run_startcmds_or_die(Connection, StartCmds),
            %%ok = give_manager_control(Sock),
            Connection;
        {error, Reason} ->
             %-% io:format("~p open connection: ... ERROR ~p~n", [self(), Reason]),
             %-% io:format("~p open connection: ... exit with failed_to_connect_to_database~n", [self()]),
            exit({failed_to_connect_to_database, Reason})
    end.

handshake(Sock, User, Password) ->
   case emysql_auth:handshake(Sock, User, Password) of
       {ok, #greeting{} = G} -> G;
       {error, Reason} ->
           gen_tcp:close(Sock),
           exit(Reason)
   end.

%% give_manager_control(Socket) ->
%%     case emysql_conn_mgr:give_manager_control(Socket) of
%%         {error, Reason} ->
%%             gen_tcp:close(Socket),
%%             exit({Reason,
%%                  "Failed to find conn mgr when opening connection. Make sure crypto is started and emysql.app is in the Erlang path."});
%%         ok -> ok
%%    end.

set_database_or_die(#emysql_connection { socket = Socket } = Connection, Database) ->
    case set_database(Connection, Database) of
        ok -> ok;
        OK1 when is_record(OK1, ok_packet) -> ok;
        Err1 when is_record(Err1, error_packet) ->
             gen_tcp:close(Socket),
             exit({failed_to_set_database, Err1#error_packet.msg})
    end.

run_startcmds_or_die(#emysql_connection{socket=Socket}, StartCmds) ->
    lists:foreach(
        fun(Cmd) ->
                Packet = <<?COM_QUERY, Cmd/binary>>,
                case emysql_tcp:send_and_recv_packet(Socket, Packet, 0) of
                    OK when OK =:= ok orelse is_record(OK, ok_packet) ->
                        ok;
                    #error_packet{msg=Msg} ->
                        gen_tcp:close(Socket),
                        exit({failed_to_run_cmd, Msg})
                end
        end,
        StartCmds
    ).
 
set_encoding_or_die(#emysql_connection { socket = Socket } = Connection, Encoding) ->
    case set_encoding(Connection, Encoding) of
        ok -> ok;
        OK2 when is_record(OK2, ok_packet) -> ok;
        Err2 when is_record(Err2, error_packet) ->
            gen_tcp:close(Socket),
            exit({failed_to_set_encoding, Err2#error_packet.msg})
    end.
 
reset_connection(PoolServer, Pools, Conn, StayLocked) ->
    %% if a process dies or times out while doing work
    %% the socket must be closed and the connection reset
    %% in the conn_mgr state. Also a new connection needs
    %% to be opened to replace the old one. If that fails,
    %% we queue the old as available for the next try
    %% by the next caller process coming along. So the
    %% pool can't run dry, even though it can freeze.
    %-% io:format("resetting connection~n"),
    MonitorRef = Conn#emysql_connection.monitor_ref,
    close_connection(Conn),
    %% OPEN NEW SOCKET
    case emysql_conn_mgr:find_pool(Conn#emysql_connection.pool_id, Pools) of
        {Pool, _} ->
            case catch open_connection(Pool) of
                #emysql_connection{} = NewConn when StayLocked == pass ->
                    NewConn2 = add_monitor_ref(NewConn, MonitorRef),
                    ok = emysql_conn_mgr:replace_connection_as_available(PoolServer, Conn, NewConn2),
                    NewConn2;
                #emysql_connection{} = NewConn when StayLocked == keep ->
                    NewConn2 = add_monitor_ref(NewConn, MonitorRef),
                    ok = emysql_conn_mgr:replace_connection_as_locked(PoolServer, Conn, NewConn2),
                    NewConn2;
                {'EXIT', Reason} ->
                    DeadConn = Conn#emysql_connection { alive = false, last_test_time = 0 },
                    emysql_conn_mgr:replace_connection_as_available(PoolServer, Conn, DeadConn),
                    {error, {cannot_reopen_in_reset, Reason}}
            end;
        undefined ->
            exit(pool_not_found)
    end.

add_monitor_ref(Conn, MonitorRef) ->
    Conn#emysql_connection{monitor_ref = MonitorRef}.

close_connections(Pool) ->
    AllConns = lists:append(
                 queue:to_list(Pool#pool.available),
                 gb_trees:values(Pool#pool.locked)
                ),
    lists:foreach(fun emysql_conn:close_connection/1, AllConns).

close_connection(Conn) ->
	%% garbage collect statements
	emysql_statements:remove(Conn#emysql_connection.id),
	ok = gen_tcp:close(Conn#emysql_connection.socket).

test_connection(PoolServer, Conn, StayLocked) ->
  case catch emysql_tcp:send_and_recv_packet(Conn#emysql_connection.socket, <<?COM_PING>>, 0) of
    {'EXIT', _} ->
      case reset_connection(PoolServer, emysql_conn_mgr:pools(PoolServer), Conn, StayLocked) of
        NewConn when is_record(NewConn, emysql_connection) ->
          NewConn;
        {error, FailedReset} ->
          exit({connection_down, {and_conn_reset_failed, FailedReset}})
      end;
      #error_packet{code = 1053,msg = _} ->

          case reset_connection(PoolServer, emysql_conn_mgr:pools(PoolServer), Conn, StayLocked) of
              NewConn when is_record(NewConn, emysql_connection) ->
                  NewConn;
              {error, FailedReset} ->
                  exit({connection_down, {and_conn_reset_failed, FailedReset}})
          end;
      _ ->
       NewConn = Conn#emysql_connection{last_test_time = now_seconds()},
       case StayLocked of
         pass -> emysql_conn_mgr:replace_connection_as_available(PoolServer, Conn, NewConn);
         keep -> emysql_conn_mgr:replace_connection_as_locked(PoolServer, Conn, NewConn)
       end,
       NewConn
  end.

need_test_connection(Conn) ->
   (Conn#emysql_connection.test_period =:= 0) orelse
     (Conn#emysql_connection.last_test_time =:= 0) orelse
     (Conn#emysql_connection.last_test_time + Conn#emysql_connection.test_period < now_seconds()).

-ifdef(timestamp_support).
now_seconds() -> os:system_time(1).
-else.
now_seconds() ->
   {M, S, _} = os:timestamp(),
   M * 1000000 + S.
-endif.

%%--------------------------------------------------------------------
%%% Internal functions
%%--------------------------------------------------------------------

%% @doc A wrapper for emysql_tcp:send_and_recv_packet/3 that may log warnings if any.
send_recv(#emysql_connection{socket = Socket, warnings = Warnings}, Packet) ->
    Ret = emysql_tcp:send_and_recv_packet(Socket, Packet, 0),
    Warnings andalso log_warnings(Socket, Ret),
    Ret.

log_warnings(Socket, #ok_packet{warning_count = WarningCount}) when WarningCount > 0 ->
    %% Fetch the warnings and log them in the OTP way.
    #result_packet{rows = WarningRows} =
        emysql_tcp:send_and_recv_packet(Socket, <<?COM_QUERY, "SHOW WARNINGS">>, 0),
    WarningMessages = [Message || [_Level, _Code, Message] <- WarningRows],
    error_logger:warning_report({emysql_warnings, WarningMessages});
log_warnings(_Sock, _OtherPacket) ->
    ok.

set_params(_, _, [], Result) -> Result;
set_params(Connection, Num, Values, _) ->
	Packet = set_params_packet(Num, Values),
	emysql_tcp:send_and_recv_packet(Connection#emysql_connection.socket, Packet, 0).

set_params_packet(NumStart, Values) ->
	BinValues = [encode(Val, binary) || Val <- Values],
	BinNums = [encode(Num, binary) || Num <- lists:seq(NumStart, NumStart + length(Values) - 1)],
	BinPairs = lists:zip(BinNums, BinValues),
	Parts = [<<"@", NumBin/binary, "=", ValBin/binary>> || {NumBin, ValBin} <- BinPairs], 
	Sets = list_to_binary(join(Parts, <<",">>)),
	<<?COM_QUERY, "SET ", Sets/binary>>.

%% @doc Join elements of list with Sep
%%
%% 1> join([1,2,3], 0).
%% [1,0,2,0,3]

join([], _Sep) -> [];
join(L, Sep) -> join(L, Sep, []).

join([H], _Sep, Acc)  -> lists:reverse([H|Acc]);
join([H|T], Sep, Acc) -> join(T, Sep, [Sep, H|Acc]).

prepare_statement(Connection, StmtName) ->
    case emysql_statements:fetch(StmtName) of
        undefined ->
            exit(statement_has_not_been_prepared);
        {Version, Statement} ->
            case emysql_statements:version(Connection#emysql_connection.id, StmtName) of
                Version ->
                    ok;
                _ ->
                    ok = prepare(Connection, StmtName, Statement),
                    emysql_statements:prepare(Connection#emysql_connection.id, StmtName, Version)
            end
    end.

% human readable string rep of the server state flag
%% @private
hstate(State) ->
       case (State band ?SERVER_STATUS_AUTOCOMMIT)   of 0 -> ""; _-> "AUTOCOMMIT " end
    ++ case (State band ?SERVER_MORE_RESULTS_EXIST)  of 0 -> ""; _-> "MORE_RESULTS_EXIST " end
    ++ case (State band ?SERVER_QUERY_NO_INDEX_USED) of 0 -> ""; _-> "NO_INDEX_USED " end.

%% @doc Encode a value so that it can be included safely in a MySQL query.
%% @spec encode(term(), list | binary) -> string() | binary() | {error, Error}
encode(null, list) ->
    "null";
encode(undefined, list) ->
    "null";
encode(null, binary)  ->
    <<"null">>;
encode(undefined, binary)  ->
    <<"null">>;
encode(Val, list) when is_binary(Val) ->
    quote(binary_to_list(Val));
encode(Val, binary) when is_atom(Val) ->
    encode(atom_to_list(Val), binary);
encode(Val, binary) when is_list(Val) ->
    list_to_binary(quote(binary_to_list(binary:list_to_bin(Val))));
encode(Val, binary) when is_binary(Val) ->
    list_to_binary(quote(binary_to_list(Val)));
encode(Val, list) when is_list(Val) ->
    quote(binary_to_list(binary:list_to_bin(Val)));
encode(Val, list) when is_integer(Val) ->
    integer_to_list(Val);
encode(Val, binary) when is_integer(Val) ->
    list_to_binary(integer_to_list(Val));
encode(Val, list) when is_float(Val) ->
    [Res] = io_lib:format("~w", [Val]),
    Res;
encode(Val, binary) when is_float(Val) ->
    iolist_to_binary(io_lib:format("~w", [Val]));
encode({datetime, Val}, ReturnType) ->
    encode(Val, ReturnType);
encode({date, Val}, ReturnType) ->
    encode(Val, ReturnType);
encode({time, Val}, ReturnType) ->
    encode(Val, ReturnType);
encode({{Year, Month, Day}, {Hour, Minute, Second}}, list) ->
    Res = io_lib:format("'~4.4.0w-~2.2.0w-~2.2.0w ~2.2.0w:~2.2.0w:~2.2.0w'",
                        [Year, Month, Day, Hour, Minute, Second]),
    lists:flatten(Res);
encode({{_Year, _Month, _Day}, {_Hour, _Minute, _Second}}=Val, binary) ->
    list_to_binary(encode(Val, list));
encode({Time1, Time2, Time3}, list) ->
    Res = two_digits([Time1, Time2, Time3]),
    lists:flatten(Res);
encode({_Time1, _Time2, _Time3}=Val, binary) ->
    list_to_binary(encode(Val, list));
encode(Val, _) ->
    {error, {unrecognized_value, Val}}.
    
%% @private
two_digits(Nums) when is_list(Nums) ->
    [two_digits(Num) || Num <- Nums];
two_digits(Num) ->
    [Str] = io_lib:format("~b", [Num]),
    case length(Str) of
        1 -> [$0 | Str];
        _ -> Str
    end.

%% @doc Quote a string or binary value so that it can be included safely in a
%% MySQL query. For the quoting, a binary is converted to a list and back.
%% For this, it's necessary to know the encoding of the binary.
%% @spec quote(x()) -> x()
%%       x() = list() | binary()
%% @end
%% hd/11,12
quote(String) when is_list(String) ->
    [39 | lists:reverse([39 | quote_loop(String)])]. %% 39 is $'

%% @doc  Make MySQL-safe backslash escapes before 10, 13, \, 26, 34, 39.
%% @spec quote_loop(list()) -> list()
%% @private
%% @end
%% hd/11,12
quote_loop(List) ->
    quote_loop(List, []).

quote_loop([], Acc) ->
    Acc;

quote_loop([0 | Rest], Acc) ->
    quote_loop(Rest, [$0, $\\ | Acc]);

quote_loop([10 | Rest], Acc) ->
    quote_loop(Rest, [$n, $\\ | Acc]);

quote_loop([13 | Rest], Acc) ->
    quote_loop(Rest, [$r, $\\ | Acc]);

quote_loop([$\\ | Rest], Acc) ->
    quote_loop(Rest, [$\\ , $\\ | Acc]);

quote_loop([39 | Rest], Acc) -> %% 39 is $'
    quote_loop(Rest, [39, $\\ | Acc]); %% 39 is $'

quote_loop([34 | Rest], Acc) -> %% 34 is $"
    quote_loop(Rest, [34, $\\ | Acc]); %% 34 is $"

quote_loop([26 | Rest], Acc) ->
    quote_loop(Rest, [$Z, $\\ | Acc]);

quote_loop([C | Rest], Acc) ->
    quote_loop(Rest, [C | Acc]).
